### PR TITLE
HDDS-8702. Extract handler for reconfigurable properties

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Some logic related to reconfiguration can be refactored into a common handler object to reduce the complexity and avoid duplication.

https://issues.apache.org/jira/browse/HDDS-8702

## How was this patch tested?

Added unit test, refactored existing integration test.

Tested CLI commands in `ozone` compose environment.

```
$ docker-compose exec om bash
bash-4.2$ ozone admin reconfig --address om:9862 properties
OM: Node [om:9862] Reconfigurable properties:
ozone.administrators
ozone.readonly.administrators
bash-4.2$ vi /etc/hadoop/ozone-site.xml
bash-4.2$ ozone admin reconfig --address om:9862 start
OM: Started reconfiguration task on node [om:9862].
bash-4.2$ ozone admin reconfig --address om:9862 status
OM: Reconfiguring status for node [om:9862]: started at Thu May 25 15:26:05 UTC 2023 and finished at Thu May 25 15:26:05 UTC 2023.
SUCCESS: Changed property ozone.administrators
	From: ""
	To: "foo"
bash-4.2$ exit

$ docker-compose exec scm bash                   
bash-4.2$ ozone admin reconfig --address scm:9860 properties
SCM: Node [scm:9860] Reconfigurable properties:
ozone.administrators
bash-4.2$ vi /etc/hadoop/ozone-site.xml
bash-4.2$ ozone admin reconfig --address scm:9860 start
SCM: Started reconfiguration task on node [scm:9860].
bash-4.2$ ozone admin reconfig --address scm:9860 status
SCM: Reconfiguring status for node [scm:9860]: started at Thu May 25 15:33:59 UTC 2023 and finished at Thu May 25 15:33:59 UTC 2023.
SUCCESS: Changed property ozone.administrators
	From: ""
	To: "bar"
bash-4.2$ exit
```

https://github.com/adoroszlai/hadoop-ozone/actions/runs/5117942592

### CODE REVIEW

**Comment by adoroszlai** [comment]
@whbing @xichen01 please take a look

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java [hunk 1]
```diff
@@ -1306,20 +1311,20 @@ public String getServerName() throws IOException {
 
   @Override
   public void startReconfigure() throws IOException {
```
Comment (xichen01) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:1313]: Maybe all OM/SCM/DN's `startReconfigure/getReconfigureStatus/listReconfigureProperties` can be implemented at `Class ReconfigurationHandler`

**Review by xichen01 — COMMENTED** [review]
LGTM. Just a little commented you can refer to

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/conf/ReconfigurationHandler.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  @Override
+  public List<String> getReconfigurableProperties() {
+    return new ArrayList<>(new TreeSet<>(properties.keySet()));
+  }
+
+  @Override
+  public String reconfigurePropertyImpl(String property, String newValue) {
+    return properties.getOrDefault(property, identity())
+        .apply(newValue);
```
[the remark this answers is not in the document]
Comment (whbing) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/conf/ReconfigurationHandler.java:64]: > maybe we only return the newValue String when the reconfig is execute normal, If the request `property` is not contain in the `properties`, we can throw an exception.

Now do nothing if the request property is not contain in the properties. We can see the result by exe `status`.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java (hunk 1 above)
Comment (whbing) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:1313]: Thanks @adoroszlai  for providing an elegant refactoring. I tested the PR in compose environment again and the result is as expected.
LGTM. +1

**Comment by whbing** [comment]
Thanks @adoroszlai  for providing an elegant refactoring. I tested the PR in compose environment again and the result is as expected.
LGTM. +1

**Comment by adoroszlai** [comment]
Thanks @whbing, @xichen01 for the review.  I have updated the patch based on your suggestion (which seems to have been deleted) to also implement `ReconfigureProtocol` in the handler.  Good idea, it helps further reduce duplication.  We can also get rid of multiple copies of `getRemoteUser()` function.

CI is still running in my fork, will mark this ready for review if it succeeds.

**Review by whbing — APPROVED** [review]

**Comment by whbing** [comment]
+1

**Comment by adoroszlai** [comment]
@captainzmc @dombizita please review

**Review by captainzmc — APPROVED** [review]
Thanks @adoroszlai for the patch. The refactoring looks good.  Let's merge this.

**Comment by adoroszlai** [comment]
Thanks @captainzmc, @whbing, @xichen01 for the review.

**Pull request merged**